### ISSUE TITLE
Incorrect return code from `hexRing` on input pentagon

### ISSUE THREAD

**Issue by ajfriend** [uber/h3#252]
The docstring of the `hexRing()` function says that it should `return 0 if no pentagonal distortion was encountered`. However, due to the "short-circuit" logic at the start of the function, it will always return `0` when `k==0`, no matter if the input `origin` is a pentagon or not:

https://github.com/uber/h3/blob/a6cd31b73aa26f0f7e4f5e5447b2c3bca20b17d6/src/h3lib/lib/algos.c#L549-L554

Do we want this behavior, or would it be better to return `HEX_RANGE_PENTAGON` in this case?

The fix should be easy, and only involve moving this check

https://github.com/uber/h3/blob/a6cd31b73aa26f0f7e4f5e5447b2c3bca20b17d6/src/h3lib/lib/algos.c#L560-L563

to run before the "short-circuit" logic.

**Comment by dfellis** [comment]
Technically no pentagonal distortion is encountered in that case, since it never traverses from the pentagon to a neighbor.

**Comment by ajfriend** [comment]
I was interpreting "pentagonal distortion" to mean "the algorithm ran into a pentagon at any point". Is that incorrect? What's the technical definition of pentagonal distortion?

**Comment by ajfriend** [comment]
And actually, beyond technical correctness, I think most people would default to assuming the same interpretation. I wonder if fixing this would be more intuitive/safer.

**Comment by dfellis** [comment]
> I was interpreting "pentagonal distortion" to mean "the algorithm ran into a pentagon at any point". Is that incorrect? What's the technical definition of pentagonal distortion?

A traversal involving a pentagon occurred, changing the number of valid IJK directions out of the pentagon, breaking the algorithm.

I don't think intentionally breaking the endpoint further is what users would want -- it's a solvable problem to correct the pentagonal distortion just as it is for crossing icosahedron faces, which also was originally broken for hexRing before H3 was open sourced. Basically, these are the limitations of the functions *right now* but we should strive for correctness *and* speed, so making this less correct seems like a step backwards.

(Side note: I believe there are 2 categories of pentagons: the two "polar" pentagons and the 10 "equatorial" pentagons, where the polar pentagons consist of the I direction of each face pointing to the center of the pentagon, while the equatorial pentagons have faces alternating between J and K pointing at the center of the pentagon. Figuring out which kind you're on and which face you're originating from should allow us to adjust the rotations appropriately for hexRing, but I don't really have the time anymore to focus on this problem.)

**Comment by dfellis** [comment]
Essentially, I see the logic end-users are after as either:

1. I will use the faster `hex_ring` algo, but any time it fails I will fall back to the `k_ring1 - k_ring2` version so I get all of the data I'm after.
2. I want as much of the data as I can possibly get for hexagons I am processing, but I am unwilling to take the slowdown when the algorithm needs to fall back (some sort of stream processing situation) and will just drop those records instead (but if you can give me more data, please do so).

**Comment by sahrk** [comment]
I agree with @dfellis on the semantics of returning 0 for success; a ring of cells with k == 0 is always well-defined even if the center cell is a pentagon. But I can see why @ajfriend is confused by some of the naming. A ring of cells of radius 0 about a pentagon is not a "hex" ring. For this reason I've been trying to consciously use "cell" rather than "hex" in my writing and code, whenever the possibility of a pentagon exists (though without looking I'll bet I broke that rule somewhere in H3). Maybe in this case the solution would be to rename `HEX_RANGE_PENTAGON` to something like `PENTAGON_DISTORTION_ERROR`.

**Comment by ajfriend** [comment]
The end-user logic @dfellis outlined totally makes sense to me, but I still think there's potential for user confusion with the docstring as-is. "Pentagonal distortion" is a technical-enough term that end users might interpret it the other way (as I did).

Also interesting: The history of the function with crossing icosahedron faces, and that the "pentagonal distortion" issue might also be solved in the future. I wonder if the docstring might reflect any of that.

Instead of `return 0 if no pentagonal distortion was encountered`, I wonder if it would be worthwhile to change the comment to something like 

```
return 0 if computation was successful
return nonzero if unsuccessful. This may happen, for example,
if a pentagon is encountered during traversal. Failure cases may
be fixed in future versions.
```

**Comment by ajfriend** [comment]
Resolved with #256
